From acd6674f182475cf8fc5c00e5011c1ff2203b927 Mon Sep 17 00:00:00 2001
From: Anton Bangratz <anton.bangratz@gmail.com>
Date: Tue, 14 Sep 2010 15:48:10 +0200
Subject: [PATCH] Human attribute names for nested labels

Labels try to look up the text in the I18n backend via the name
of the object the attributes are nested for.
---
 actionpack/lib/action_view/helpers/form_helper.rb |    7 ++++-
 actionpack/test/template/form_helper_test.rb      |   27 +++++++++++++++++++++
 2 files changed, 33 insertions(+), 1 deletions(-)

diff --git a/actionpack/lib/action_view/helpers/form_helper.rb b/actionpack/lib/action_view/helpers/form_helper.rb
index 79a9d99..ac620ad 100644
--- a/actionpack/lib/action_view/helpers/form_helper.rb
+++ b/actionpack/lib/action_view/helpers/form_helper.rb
@@ -892,7 +892,12 @@ module ActionView
           label_tag(name_and_id["id"], options, &block)
         else
           content = if text.blank?
-            I18n.t("helpers.label.#{object_name}.#{method_name}", :default => "").presence
+            if object_name =~ /.+\[(.+)_attributes\]/ then
+              base_name = $1
+              I18n.t("helpers.label.#{base_name}.#{method_name}", :default => "").presence
+            else
+              I18n.t("helpers.label.#{object_name}.#{method_name}", :default => "").presence
+            end
           else
             text.to_s
           end
diff --git a/actionpack/test/template/form_helper_test.rb b/actionpack/test/template/form_helper_test.rb
index 97a08d4..4dde9c4 100644
--- a/actionpack/test/template/form_helper_test.rb
+++ b/actionpack/test/template/form_helper_test.rb
@@ -55,6 +55,17 @@ class FormHelperTest < ActionView::TestCase
       }
     }
 
+    # Create "nested" locale for testing I18n label helpers
+    I18n.backend.store_translations 'nested', {
+      :helpers => {
+        :label => {
+          :author => {
+            :name => "Author Name"
+          }
+        }
+      }
+    }
+
     @post = Post.new
     @comment = Comment.new
     def @post.errors()
@@ -966,6 +977,22 @@ class FormHelperTest < ActionView::TestCase
     assert_dom_equal expected, output_buffer
   end
 
+  def test_form_for_with_nested_fields_for_and_humanized_label_text
+    @post.author = Author.new
+    old_locale, I18n.locale = I18n.locale, :nested
+    output_buffer = form_for(@post) do |f|
+      concat f.fields_for(:author) { |c|
+        concat c.label(:name)
+      }
+    end
+    expected =  whole_form('/posts/123', 'edit_post_123', 'edit_post', :method => :put) do
+                  "<label for='post_author_attributes_name'>Author Name</label>"
+                end
+    assert_dom_equal expected, output_buffer
+  ensure
+    I18n.locale = old_locale
+  end
+
   def test_nested_fields_for_with_index_on_both
     assert_deprecated do
       form_for(:post, @post, :index => 1) do |f|
-- 
1.7.1

